src/MySqlConnector/Protocol/Serialization/AuthGSSAPI.cs

```diff
@@ -0,0 +1,237 @@
+using System.Net;
+using System.Net.Security;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using MySqlConnector.Core;
+using MySqlConnector.Utilities;
+
+namespace MySqlConnector.Protocol.Serialization
+{
+	internal class NegotiateStreamConstants
+	{
+		public const int HeaderLength = 5;
+		public const byte MajorVersion = 1;
+		public const byte MinorVersion = 0;
+		public const byte HandshakeDone = 0x14;
+		public const byte HandshakeError = 0x15;
+		public const byte HandshakeInProgress = 0x16;
+		public const ushort MaxPayloadLength = ushort.MaxValue;
+	}
+
+	/// <summary>
+	/// Helper class to translate NegotiateStream framing for SPNEGO token
+	/// into MySQL protocol packets.
+	///
+	/// Serves as underlying stream for System.Net.NegotiateStream
+	/// to perform MariaDB's auth_gssapi_client authentication.
+	///
+	/// NegotiateStream protocol is described in e.g here
+	/// https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NNS/[MS-NNS].pdf
+	/// We only use Handshake Messages for authentication.
+	/// </summary>
+
+	internal class NegotiateToMySqlConverterStream : Stream
+	{
+		bool m_clientHandshakeDone;
+
+		MemoryStream m_readBuffer;
+		MemoryStream m_writeBuffer;
+		int m_writePayloadLength;
+		ServerSession m_serverSession;
+		IOBehavior m_ioBehavior;
+		CancellationToken m_cancellationToken;
+
+		public PayloadData MySQLProtocolPayload { get; private set; }
+		public NegotiateToMySqlConverterStream(ServerSession serverSession, IOBehavior ioBehavior, CancellationToken cancellationToken)
+		{
+			m_serverSession = serverSession;
+			m_readBuffer = new MemoryStream();
+			m_writeBuffer = new MemoryStream();
+			m_ioBehavior = ioBehavior;
+			m_cancellationToken = cancellationToken;
+		}
+
+		static void CreateNegotiateStreamMessageHeader(byte[] buffer, int offset, byte messageId, long payloadLength)
+		{
+			buffer[offset] =  messageId;
+			buffer[offset+1] = NegotiateStreamConstants.MajorVersion;
+			buffer[offset+2] = NegotiateStreamConstants.MinorVersion;
+			buffer[offset+3] = (byte) (payloadLength >> 8);
+			buffer[offset+4] = (byte) (payloadLength & 0xff);
+		}
+		public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+		{
+			var bytesRead = 0;
+
+			if (m_readBuffer.Length == m_readBuffer.Position)
+			{
+				if (count < NegotiateStreamConstants.HeaderLength)
+					throw new InvalidDataException("Unexpected call to read less then NegotiateStream header");
+
+				if (m_clientHandshakeDone)
+				{
+					// NegotiateStream protocol expects server to send "handshake done"
+					// empty message at the end of handshake.
+					CreateNegotiateStreamMessageHeader(buffer, offset, NegotiateStreamConstants.HandshakeDone, 0);
+					return NegotiateStreamConstants.HeaderLength;
+				}
+				// Read and cache packet from server.
+				var payload = await m_serverSession.ReceiveReplyAsync(m_ioBehavior, cancellationToken);
```

Comment:
```suggestion
				var payload = await m_serverSession.ReceiveReplyAsync(m_ioBehavior, cancellationToken).ConfigureAwait(false);
```

Should always use `ConfigureAwait(false)` to avoid potential deadlock; cf. https://twitter.com/bgrainger/status/938167570540007424

Reply:
OK, applied the fix in new commit.